File: filter_component_editor/filter_component_editor/pipeline_graph.py
# Copyright 2026 Maik Knof
# SPDX-License-Identifier: Apache-2.0

from dataclasses import dataclass, field
from typing import Any

import yaml


@dataclass
class PortRef:
    node: str
    port: str = ""
    direction: str = ""

    def to_dict(self) -> dict[str, str]:
        data = {"node": self.node, "direction": self.direction}
        if self.port:
            data["port"] = self.port
        return data


@dataclass
class Edge:
    source: PortRef
    target: PortRef
    topic: str = ""
    qos: dict[str, Any] = field(default_factory=dict)
    position: dict[str, float] = field(default_factory=dict)
    compatibility: str = ""

    def to_dict(self) -> dict[str, Any]:
        data: dict[str, Any] = {"from": self.source.to_dict(), "to": self.target.to_dict()}
        if self.topic:
            data["topic"] = self.topic
        if self.position:
            data["position"] = self.position
        if self.compatibility == "ros_message":
            data["compatibility"] = self.compatibility
        return data


@dataclass
class Node:
    id: str
    type: str
    name: str = ""
    package: str = ""
    filter: str = ""
    component_class: str = ""
    input_type: str = ""
    output_type: str = ""
    input_ports: str = ""
    output_ports: str = ""
    topic: str = ""
    parameters: dict[str, Any] = field(default_factory=dict)
    qos: dict[str, Any] = field(default_factory=dict)
    inputs: dict[str, Any] = field(default_factory=dict)
    outputs: dict[str, Any] = field(default_factory=dict)
    sync: dict[str, Any] = field(default_factory=dict)
    position: dict[str, float] = field(default_factory=lambda: {"x": 0.0, "y": 0.0})

    def to_dict(self) -> dict[str, Any]:
        data: dict[str, Any] = {
            "type": self.type,
            "position": self.position,
        }
        if self.type == "filter":
            data["name"] = self.name or self.id
        for key in (
            "name",
            "package",
            "filter",
            "component_class",
            "input_type",
            "output_type",
            "input_ports",
            "output_ports",
            "topic",
        ):
            value = getattr(self, key)
            if value:
                data[key] = value
        for key in ("parameters", "inputs", "outputs", "sync"):
            value = getattr(self, key)
            if value and (self.type == "filter" or key not in {"inputs", "outputs"}):
                data[key] = value
        return data


@dataclass
class Graph:
    version: int = 2
    editor: dict[str, Any] = field(default_factory=dict)
    nodes: list[Node] = field(default_factory=list)
    edges: list[Edge] = field(default_factory=list)

    def validate(
        self,
        type_by_node: dict[str, str] | None = None,
        message_type_by_logical: dict[str, str] | None = None,
    ) -> None:
        if self.version != 2:
            raise ValueError(f"unsupported graph version {self.version}")
        ids: set[str] = set()
        nodes_by_id: dict[str, Node] = {}
        for node in self.nodes:
            if not node.id:
                raise ValueError("node id must not be empty")
            if node.id in ids:
                raise ValueError(f"duplicate node id {node.id}")
            ids.add(node.id)
            nodes_by_id[node.id] = node
            if node.type not in {"filter", "topic"}:
                raise ValueError(f"unsupported node type {node.type}")
            if node.type == "filter" and not (node.component_class or (node.package and node.filter)):
                raise ValueError(f"filter node {node.id} has no component identity")
            if node.type == "topic" and not node.topic:
                raise ValueError(f"topic node {node.id} has no topic")
            if node.type == "topic" and not (node.input_type or node.output_type):
                raise ValueError(f"topic node {node.id} has no type")
            if node.type == "filter":
                _validate_sync_config(node)

        for edge in self.edges:
            if edge.compatibility and edge.compatibility != "ros_message":
                raise ValueError(f"unsupported edge compatibility {edge.compatibility}")
            if edge.source.direction != "output":
                raise ValueError(f"edge from.direction must be output for {edge.source.node}")
            if edge.target.direction != "input":
                raise ValueError(f"edge to.direction must be input for {edge.target.node}")
            if edge.source.node not in ids:
                raise ValueError(f"edge source {edge.source.node} does not exist")
            if edge.target.node not in ids:
                raise ValueError(f"edge target {edge.target.node} does not exist")
            if edge.source.node == edge.target.node:
                raise ValueError(f"self edge on {edge.source.node} is invalid")
            source_type = self._node_type(nodes_by_id[edge.source.node], outgoing=True, port=edge.source.port)
            target_type = self._node_type(nodes_by_id[edge.target.node], outgoing=False, port=edge.target.port)
            if type_by_node:
                source_type = type_by_node.get(edge.source.node, source_type)
                target_type = type_by_node.get(edge.target.node, target_type)
            if source_type and target_type and source_type != target_type:
                if edge.compatibility == "ros_message":
                    if message_type_by_logical is not None:
                        source_message = message_type_by_logical.get(source_type, "")
                        target_message = message_type_by_logical.get(target_type, "")
                        if not source_message or source_message != target_message:
                            raise ValueError(
                                f"type mismatch: {edge.source.node} produces {source_type}, "
                                f"{edge.target.node} expects {target_type}"
                            )
                    continue
                raise ValueError(
                    f"type mismatch: {edge.source.node} produces {source_type}, "
                    f"{edge.target.node} expects {target_type}"
                )
        used_inputs: set[tuple[str, str]] = set()
        used_outputs: set[tuple[str, str]] = set()
        for edge in self.edges:
            source_node = nodes_by_id[edge.source.node]
            target_node = nodes_by_id[edge.target.node]
            if source_node.type == "filter":
                key = (
                    source_node.id,
                    _canonical_filter_port(source_node, edge.source.port, True),
                )
                if key in used_outputs:
                    raise ValueError(f"filter output {source_node.id}:{key[1]} is already connected")
                used_outputs.add(key)
            if target_node.type == "filter":
                key = (
                    target_node.id,
                    _canonical_filter_port(target_node, edge.target.port, False),
                )
                if key in used_inputs:
                    raise ValueError(f"filter input {target_node.id}:{key[1]} is already connected")
                used_inputs.add(key)
        published_topics_by_filter: dict[str, set[str]] = {}
        subscribed_topics_by_filter: dict[str, set[str]] = {}
        for edge in self.edges:
            source_node = nodes_by_id[edge.source.node]
            target_node = nodes_by_id[edge.target.node]
            if source_node.type == "filter" and target_node.type == "topic":
                topic = target_node.topic or target_node.id
                published_topics_by_filter.setdefault(source_node.id, set()).add(topic)
            elif source_node.type == "topic" and target_node.type == "filter":
                topic = source_node.topic or source_node.id
                subscribed_topics_by_filter.setdefault(target_node.id, set()).add(topic)
        for filter_id, published_topics in published_topics_by_filter.items():
            repeated_topics = published_topics & subscribed_topics_by_filter.get(filter_id, set())
            if repeated_topics:
                topic = sorted(repeated_topics)[0]
                raise ValueError(f"filter {filter_id} cannot both publish and subscribe to topic {topic}")
        topic_names = [node.topic for node in self.nodes if node.type == "topic" and node.topic]
        if len(topic_names) != len(set(topic_names)):
            raise ValueError("topic names must be unique")

    @staticmethod
    def _node_type(node: Node, outgoing: bool, port: str = "") -> str:
        if node.type == "topic":
            return _type_for_port(node.output_type or node.input_type, port, outgoing)
        return _type_for_port(
            (node.output_ports or node.output_type) if outgoing else (node.input_ports or node.input_type),
            port,
            outgoing,
        )

    def to_dict(self) -> dict[str, Any]:
        self._canonicalize_edge_ports()
        return {
            "version": self.version,
            "editor": self.editor,
            "nodes": [node.to_dict() for node in self.nodes],
            "edges": [edge.to_dict() for edge in self.edges],
        }

    def _canonicalize_edge_ports(self) -> None:
        nodes_by_id = {node.id: node for node in self.nodes}
        for edge in self.edges:
            source = nodes_by_id.get(edge.source.node)
            target = nodes_by_id.get(edge.target.node)
            if source is not None and source.type == "filter":
                edge.source.port = _canonical_filter_port(source, edge.source.port, True)
            if target is not None and target.type == "filter":
                edge.target.port = _canonical_filter_port(target, edge.target.port, False)


def graph_from_dict(data: dict[str, Any]) -> Graph:
    graph = Graph(version=int(data.get("version", 0)))
    graph.editor = data.get("editor", {}) or {}
    for item in data.get("nodes", []):
        node_type = item["type"]
        node_id = item.get("id", item.get("name", item.get("topic", "")))
        parameters = item.get("parameters", {}) or {}
        sync = item.get("sync", {}) or {}
        if node_type == "filter":
            parameters = dict(parameters)
            sync = dict(sync)
            for key in ("policy", "queue_size", "slop", "max_interval"):
                if key in parameters:
                    sync.setdefault(key, parameters.pop(key))
        graph.nodes.append(
            Node(
                id=node_id,
                type=node_type,
                name=item.get("name", ""),
                package=item.get("package", ""),
                filter=item.get("filter", ""),
                component_class=item.get("component_class", ""),
                input_type=item.get("input_type", ""),
                output_type=item.get("output_type", ""),
                input_ports=item.get("input_ports", ""),
                output_ports=item.get("output_ports", ""),
                topic=item.get("topic", ""),
                parameters=parameters,
                qos={},
                inputs=item.get("inputs", {}) or {},
                outputs=item.get("outputs", {}) or {},
                sync=sync,
                position=item.get("position", {"x": 0.0, "y": 0.0}) or {"x": 0.0, "y": 0.0},
            )
        )
    for item in data.get("edges", []):
        graph.edges.append(
            Edge(
                PortRef(item["from"]["node"], item["from"].get("port", ""), item["from"].get("direction", "")),
                PortRef(item["to"]["node"], item["to"].get("port", ""), item["to"].get("direction", "")),
                item.get("topic", ""),
                {},
                item.get("position", {}) or {},
                item.get("compatibility", ""),
            )
        )
    graph.validate()
    return graph


def load_graph(path: str) -> Graph:
    with open(path, "r", encoding="utf-8") as stream:
        return graph_from_dict(yaml.safe_load(stream) or {})


def save_graph(graph: Graph, path: str) -> None:
    graph.validate()
    with open(path, "w", encoding="utf-8") as stream:
        yaml.safe_dump(graph.to_dict(), stream, sort_keys=False)


def _split_types(value: str) -> list[str]:
    return [item.strip() for item in value.replace(";", ",").split(",") if item.strip()]


def _split_ports(value: str) -> list[tuple[str, str]]:
    ports: list[tuple[str, str]] = []
    for item in _split_types(value):
        if ":" in item:
            name, stream_type = item.split(":", 1)
            ports.append((name.strip(), stream_type.strip()))
        else:
            ports.append(("", item))
    return [(name, stream_type) for name, stream_type in ports if stream_type]


def _port_name_for_type(stream_type: str, index: int, total: int, outgoing: bool) -> str:
    if total > 1 and not outgoing:
        return f"input_{index + 1}"
    if stream_type == "PointIndices":
        return "indices"
    if stream_type.startswith("Point"):
        return "cloud"
    return stream_type.replace("/", "_").replace(":", "_").lower() or "out"


def _type_for_port(value: str, port: str, outgoing: bool) -> str:
    ports = _split_ports(value)
    if not ports:
        return ""
    if not port or port in {"in", "out"}:
        return ports[0][1]
    for index, (port_name, stream_type) in enumerate(ports):
        inferred_port = _port_name_for_type(stream_type, index, len(ports), outgoing)
        if port == stream_type or port == port_name or port == inferred_port:
            return stream_type
    return ports[0][1] if outgoing else ""


def _canonical_port(value: str, port: str, outgoing: bool) -> str:
    ports = _split_ports(value)
    default_port = "out" if outgoing else "in"
    if not ports:
        return port or default_port
    if not port or port in {"in", "out"}:
        port_name, stream_type = ports[0]
        if port_name:
            return port_name
        return _port_name_for_type(stream_type, 0, len(ports), outgoing) if len(ports) > 1 else default_port
    valid_ports = {
        port_name or _port_name_for_type(stream_type, index, len(ports), outgoing)
        for index, (port_name, stream_type) in enumerate(ports)
    }
    return port if port in valid_ports else port


def _canonical_filter_port(node: Node, port: str, outgoing: bool) -> str:
    spec = (node.output_ports or node.output_type) if outgoing else (node.input_ports or node.input_type)
    canonical = _canonical_port(spec, port, outgoing)
    default_port = "out" if outgoing else "in"
    configs = node.outputs if outgoing else node.inputs
    if canonical == default_port and (not port or port in {"in", "out"}) and len(configs) == 1:
        return next(iter(configs))
    return canonical


def _validate_sync_config(node: Node) -> None:
    for key, value in node.sync.items():
        if key == "policy":
            raise ValueError(f"filter node {node.id} uses removed sync.policy; use sync.mode instead")
        if key == "slop":
            raise ValueError(f"filter node {node.id} uses removed sync.slop; use sync.max_interval instead")
        if key not in {"mode", "queue_size", "max_interval"}:
            raise ValueError(f"unsupported sync option {key} on filter node {node.id}")
        if key == "mode" and value not in {"receipt_time", "latest"}:
            raise ValueError(f"unsupported sync.mode {value} on filter node {node.id}")

File: filter_component_factory/src/pipeline_graph.cpp
// Copyright 2026 Maik Knof
// SPDX-License-Identifier: Apache-2.0

#include "filter_component_factory/pipeline/pipeline_graph.hpp"

#include <algorithm>
#include <cctype>
#include <set>
#include <stdexcept>
#include <string>
#include <sstream>
#include <utility>
#include <vector>

#include <yaml-cpp/yaml.h>

namespace filter_component_factory::pipeline
{

namespace
{

std::string requireString(const YAML::Node & node, const std::string & key)
{
  if (!node[key] || !node[key].IsScalar()) {
    throw std::runtime_error("Missing string field '" + key + "'");
  }
  return node[key].as<std::string>();
}

std::string optionalString(const YAML::Node & node, const std::string & key)
{
  if (!node[key]) {
    return {};
  }
  return node[key].as<std::string>();
}

std::map<std::string, std::string> stringMap(const YAML::Node & node)
{
  std::map<std::string, std::string> values;
  if (!node) {
    return values;
  }
  if (!node.IsMap()) {
    throw std::runtime_error("Expected a mapping");
  }
  for (const auto & entry : node) {
    if (entry.second.IsScalar()) {
      values.emplace(entry.first.as<std::string>(), entry.second.as<std::string>());
    } else {
      values.emplace(entry.first.as<std::string>(), YAML::Dump(entry.second));
    }
  }
  return values;
}

std::map<std::string, std::map<std::string, std::string>> portConfigMap(const YAML::Node & node)
{
  std::map<std::string, std::map<std::string, std::string>> values;
  if (!node) {
    return values;
  }
  if (!node.IsMap()) {
    throw std::runtime_error("Expected a port configuration mapping");
  }
  for (const auto & entry : node) {
    const auto port_name = entry.first.as<std::string>();
    const auto qos = entry.second["qos"] ? stringMap(entry.second["qos"]) :
      std::map<std::string, std::string>{};
    values.emplace(port_name, qos);
  }
  return values;
}

PipelinePort parsePort(const YAML::Node & node)
{
  if (!node || !node.IsMap()) {
    throw std::runtime_error("Edge endpoint must be a mapping");
  }
  PipelinePort port;
  port.node = requireString(node, "node");
  port.port = optionalString(node, "port");
  port.direction = requireString(node, "direction");
  return port;
}

std::vector<std::string> splitTypes(const std::string & value)
{
  std::vector<std::string> types;
  std::stringstream stream{value};
  std::string item;
  while (std::getline(stream, item, ',')) {
    const auto first = item.find_first_not_of(" \t\n\r");
    if (first == std::string::npos) {
      continue;
    }
    const auto last = item.find_last_not_of(" \t\n\r");
    types.push_back(item.substr(first, last - first + 1U));
  }
  return types;
}

std::vector<std::pair<std::string, std::string>> splitPorts(const std::string & value)
{
  std::vector<std::pair<std::string, std::string>> ports;
  for (const auto & item : splitTypes(value)) {
    const auto separator = item.find(':');
    if (separator == std::string::npos) {
      ports.push_back({"", item});
      continue;
    }
    auto name = item.substr(0, separator);
    auto stream_type = item.substr(separator + 1U);
    const auto name_first = name.find_first_not_of(" \t\n\r");
    const auto name_last = name.find_last_not_of(" \t\n\r");
    name = name_first == std::string::npos ? std::string{} :
      name.substr(name_first, name_last - name_first + 1U);
    const auto type_first = stream_type.find_first_not_of(" \t\n\r");
    const auto type_last = stream_type.find_last_not_of(" \t\n\r");
    stream_type = type_first == std::string::npos ? std::string{} :
      stream_type.substr(type_first, type_last - type_first + 1U);
    if (!stream_type.empty()) {
      ports.push_back({name, stream_type});
    }
  }
  return ports;
}

std::string portNameForType(
  const std::string & stream_type,
  size_t index,
  size_t total,
  bool outgoing)
{
  if (!outgoing && total > 1U) {
    return "input_" + std::to_string(index + 1U);
  }
  if (stream_type == "PointIndices") {
    return "indices";
  }
  if (stream_type.rfind("Point", 0) == 0) {
    return "cloud";
  }
  auto name = stream_type;
  std::replace(name.begin(), name.end(), '/', '_');
  std::replace(name.begin(), name.end(), ':', '_');
  std::transform(name.begin(), name.end(), name.begin(), [](unsigned char value) {
      return static_cast<char>(std::tolower(value));
    });
  return name.empty() ? "out" : name;
}

std::string typeForTopicPort(const std::string & value, const std::string & port, bool outgoing)
{
  const auto ports = splitPorts(value);
  if (ports.empty()) {
    return {};
  }
  if (port.empty() || port == "in" || port == "out") {
    return ports.front().second;
  }
  for (size_t index = 0; index < ports.size(); ++index) {
    const auto & port_name = ports[index].first;
    const auto & stream_type = ports[index].second;
    if (
      port == stream_type ||
      port == port_name ||
      port == portNameForType(stream_type, index, ports.size(), outgoing))
    {
      return stream_type;
    }
  }
  return outgoing ? ports.front().second : std::string{};
}

std::string typeForFilterPort(const std::string & value, const std::string & port, bool outgoing)
{
  const auto ports = splitPorts(value);
  if (ports.empty() || port.empty() || port == "in" || port == "out") {
    return {};
  }
  for (size_t index = 0; index < ports.size(); ++index) {
    const auto & port_name = ports[index].first;
    const auto & stream_type = ports[index].second;
    if (
      port == stream_type ||
      port == port_name ||
      port == portNameForType(stream_type, index, ports.size(), outgoing))
    {
      return stream_type;
    }
  }
  return {};
}

bool filterPortIsValid(const PipelineNode & node, const std::string & port, bool outgoing)
{
  const auto spec = outgoing ?
    (node.output_ports.empty() ? node.output_type : node.output_ports) :
    (node.input_ports.empty() ? node.input_type : node.input_ports);
  return !typeForFilterPort(spec, port, outgoing).empty();
}

std::string canonicalPort(const std::string & value, const std::string & port, bool outgoing)
{
  const auto ports = splitPorts(value);
  const auto default_port = outgoing ? std::string{"out"} : std::string{"in"};
  if (ports.empty()) {
    return port.empty() ? default_port : port;
  }
  if (port.empty() || port == "in" || port == "out") {
    if (!ports.front().first.empty()) {
      return ports.front().first;
    }
    return ports.size() > 1U ? portNameForType(ports.front().second, 0U, ports.size(), outgoing) : default_port;
  }
  for (size_t index = 0; index < ports.size(); ++index) {
    const auto & port_name = ports[index].first;
    const auto inferred_port = portNameForType(ports[index].second, index, ports.size(), outgoing);
    if (port == (port_name.empty() ? inferred_port : port_name)) {
      return port;
    }
  }
  return port;
}

std::string nodeTypeForEdge(const PipelineNode & node, bool outgoing, const std::string & port)
{
  if (node.type == "topic") {
    return typeForTopicPort(node.output_type.empty() ? node.input_type : node.output_type, port, outgoing);
  }
  return typeForFilterPort(
    outgoing ?
    (node.output_ports.empty() ? node.output_type : node.output_ports) :
    (node.input_ports.empty() ? node.input_type : node.input_ports),
    port,
    outgoing);
}

void validateSyncConfig(const PipelineNode & node)
{
  for (const auto & [key, value] : node.sync) {
    (void)value;
    if (key == "policy") {
      throw std::runtime_error(
        "Filter node '" + node.id + "' uses removed sync.policy; use sync.mode instead");
    }
    if (key == "slop") {
      throw std::runtime_error(
        "Filter node '" + node.id + "' uses removed sync.slop; use sync.max_interval instead");
    }
    if (key != "mode" && key != "queue_size" && key != "max_interval") {
      throw std::runtime_error("Unsupported sync option '" + key + "' on filter node '" + node.id + "'");
    }
  }
  const auto mode = node.sync.find("mode");
  if (mode != node.sync.end() && mode->second != "receipt_time" && mode->second != "latest") {
    throw std::runtime_error(
      "Unsupported sync.mode '" + mode->second + "' on filter node '" + node.id + "'");
  }
}

}  // namespace

PipelineGraph loadPipelineGraph(const std::string & path)
{
  const auto root = YAML::LoadFile(path);
  if (!root || !root.IsMap()) {
    throw std::runtime_error("Pipeline YAML must contain a top-level mapping");
  }

  PipelineGraph graph;
  graph.version = root["version"] ? root["version"].as<int>() : 0;
  if (!root["nodes"] || !root["nodes"].IsSequence()) {
    throw std::runtime_error("Pipeline YAML must contain a nodes sequence");
  }
  if (!root["edges"] || !root["edges"].IsSequence()) {
    throw std::runtime_error("Pipeline YAML must contain an edges sequence");
  }

  for (const auto & item : root["nodes"]) {
    PipelineNode node;
    node.type = requireString(item, "type");
    node.name = optionalString(item, "name");
    node.package_name = optionalString(item, "package");
    node.filter = optionalString(item, "filter");
    node.component_class = optionalString(item, "component_class");
    node.python_module = optionalString(item, "python_module");
    node.python_class = optionalString(item, "python_class");
    node.implementation = optionalString(item, "implementation");
    if (node.implementation.empty()) {
      node.implementation =
        (!node.python_module.empty() || !node.python_class.empty()) ? "python" : "cpp";
    }
    node.input_type = optionalString(item, "input_type");
    node.output_type = optionalString(item, "output_type");
    node.input_ports = optionalString(item, "input_ports");
    node.output_ports = optionalString(item, "output_ports");
    node.shm_keys = optionalString(item, "shm_keys");
    node.topic = optionalString(item, "topic");
    if (item["id"]) {
      node.id = requireString(item, "id");
    } else if (!node.name.empty()) {
      node.id = node.name;
    } else {
      node.id = node.topic;
    }
    node.parameters = stringMap(item["parameters"]);
    node.qos = {};
    node.inputs = portConfigMap(item["inputs"]);
    node.outputs = portConfigMap(item["outputs"]);
    node.sync = stringMap(item["sync"]);
    if (item["shm"] && item["shm"].IsMap()) {
      node.shm_remappings = stringMap(item["shm"]["remappings"]);
    }
    if (item["position"] && item["position"].IsMap()) {
      node.x = item["position"]["x"] ? item["position"]["x"].as<double>() : 0.0;
      node.y = item["position"]["y"] ? item["position"]["y"].as<double>() : 0.0;
    }
    graph.nodes.push_back(std::move(node));
  }

  for (const auto & item : root["edges"]) {
    PipelineEdge edge;
    edge.from = parsePort(item["from"]);
    edge.to = parsePort(item["to"]);
    edge.topic = optionalString(item, "topic");
    edge.qos = {};
    if (item["position"] && item["position"].IsMap()) {
      edge.x = item["position"]["x"] ? item["position"]["x"].as<double>() : 0.0;
      edge.y = item["position"]["y"] ? item["position"]["y"].as<double>() : 0.0;
    }
    graph.edges.push_back(std::move(edge));
  }

  validatePipelineGraph(graph);
  return graph;
}

void validatePipelineGraph(const PipelineGraph & graph)
{
  if (graph.version != 2) {
    throw std::runtime_error("Unsupported pipeline graph version " + std::to_string(graph.version));
  }

  std::set<std::string> ids;
  std::set<std::string> topic_names;
  for (const auto & node : graph.nodes) {
    if (node.id.empty()) {
      throw std::runtime_error("Pipeline node id must not be empty");
    }
    if (!ids.insert(node.id).second) {
      throw std::runtime_error("Duplicate pipeline node id '" + node.id + "'");
    }
    if (node.type != "filter" && node.type != "topic") {
      throw std::runtime_error("Unsupported node type '" + node.type + "' on node '" + node.id + "'");
    }
    if (node.type == "filter" && node.implementation != "cpp" && node.implementation != "python") {
      throw std::runtime_error(
        "Unsupported filter implementation '" + node.implementation + "' on node '" + node.id + "'");
    }
    if (node.type == "filter" && node.implementation == "cpp" && node.component_class.empty()) {
      throw std::runtime_error("Filter node '" + node.id + "' has no component class");
    }
    if (
      node.type == "filter" && node.implementation == "python" &&
      (node.python_module.empty() || node.python_class.empty()))
    {
      throw std::runtime_error("Python filter node '" + node.id + "' has no python module or class");
    }
    if (node.type == "topic" && node.topic.empty()) {
      throw std::runtime_error("Topic node '" + node.id + "' must declare a topic");
    }
    if (node.type == "topic" && node.input_type.empty() && node.output_type.empty()) {
      throw std::runtime_error("Topic node '" + node.id + "' must declare a type");
    }
    if (node.type == "topic" && !topic_names.insert(node.topic).second) {
      throw std::runtime_error("Duplicate topic name '" + node.topic + "'");
    }
    if (node.type == "filter") {
      validateSyncConfig(node);
    }
  }

  for (const auto & edge : graph.edges) {
    if (edge.from.direction != "output") {
      throw std::runtime_error("Edge from.direction must be 'output' for node '" + edge.from.node + "'");
    }
    if (edge.to.direction != "input") {
      throw std::runtime_error("Edge to.direction must be 'input' for node '" + edge.to.node + "'");
    }
    if (ids.count(edge.from.node) == 0U) {
      throw std::runtime_error("Edge references unknown source node '" + edge.from.node + "'");
    }
    if (ids.count(edge.to.node) == 0U) {
      throw std::runtime_error("Edge references unknown target node '" + edge.to.node + "'");
    }
    if (edge.from.node == edge.to.node) {
      throw std::runtime_error("Self edges are not valid for node '" + edge.from.node + "'");
    }
    const auto source = std::find_if(
      graph.nodes.begin(),
      graph.nodes.end(),
      [&edge](const auto & node) {return node.id == edge.from.node;});
    const auto target = std::find_if(
      graph.nodes.begin(),
      graph.nodes.end(),
      [&edge](const auto & node) {return node.id == edge.to.node;});
    if (source->type == "filter" && !filterPortIsValid(*source, edge.from.port, true)) {
      throw std::runtime_error(
        "Filter output port '" + edge.from.port + "' is not declared on node '" + source->id + "'");
    }
    if (target->type == "filter" && !filterPortIsValid(*target, edge.to.port, false)) {
      throw std::runtime_error(
        "Filter input port '" + edge.to.port + "' is not declared on node '" + target->id + "'");
    }
    const auto source_type = nodeTypeForEdge(*source, true, edge.from.port);
    const auto target_type = nodeTypeForEdge(*target, false, edge.to.port);
    if (!source_type.empty() && !target_type.empty() && source_type != target_type) {
      throw std::runtime_error(
        "Type-incompatible edge from '" + edge.from.node + "' (" + source_type + ") to '" +
        edge.to.node + "' (" + target_type + ")");
    }
  }

  std::set<std::pair<std::string, std::string>> used_inputs;
  std::set<std::pair<std::string, std::string>> used_outputs;
  for (const auto & edge : graph.edges) {
    const auto source = std::find_if(
      graph.nodes.begin(),
      graph.nodes.end(),
      [&edge](const auto & node) {return node.id == edge.from.node;});
    const auto target = std::find_if(
      graph.nodes.begin(),
      graph.nodes.end(),
      [&edge](const auto & node) {return node.id == edge.to.node;});
    if (source->type == "filter") {
      const auto key = std::make_pair(
        source->id,
        canonicalPort(source->output_ports.empty() ? source->output_type : source->output_ports, edge.from.port, true));
      if (!used_outputs.insert(key).second) {
        throw std::runtime_error("Filter output '" + key.first + ":" + key.second + "' is already connected");
      }
    }
    if (target->type == "filter") {
      const auto key = std::make_pair(
        target->id,
        canonicalPort(target->input_ports.empty() ? target->input_type : target->input_ports, edge.to.port, false));
      if (!used_inputs.insert(key).second) {
        throw std::runtime_error("Filter input '" + key.first + ":" + key.second + "' is already connected");
      }
    }
  }
}

}  // namespace filter_component_factory::pipeline
